### PR TITLE
/forcepos and fixed area lock shenanigans

### PR DESCRIPTION
**Pull request by Crystalwarrior**
Adds /forcepos [pos] [target]
you can omit [target] to force everyone in area into pos [pos]
This is a CM or mod-only command
Fixed area locking and made it so you can peanut in areas but unable to speak ICly unless invited.
Added /uninvite and made /area_kick command mod-only.

### CODE REVIEW

File: server/commands.py [hunk 1]
```diff
@@ -181,6 +181,49 @@ def ooc_cmd_pos(client, arg):
         client.area.broadcast_evidence_list()
         client.send_host_message('Position changed.')   
 
+def ooc_cmd_forcepos(client, arg):
```
Comment (oldmud0) [inline server/commands.py:184]: Needs documentation in readme.

File: server/commands.py [hunk 2]
```diff
@@ -527,16 +570,41 @@ def ooc_cmd_invite(client, arg):
...
-        client.area.invite_list[client.server.client_manager.get_targets(client, TargetType.ID, int(arg), False)[0].ipid] = None
-        client.send_host_message('{} is invited to your area.'.format(client.server.client_manager.get_targets(client, TargetType.ID, int(arg), False)[0].get_char_name()))
+        c = client.server.client_manager.get_targets(client, TargetType.ID, int(arg), False)[0]
+        client.area.invite_list[c.ipid] = None
+        client.send_host_message('{} is invited to your area.'.format(c.get_char_name()))
+        c.send_host_message('You were invited and given access to area {}.'.format(client.area.id))
     except:
         raise ClientError('You must specify a target. Use /invite <id>')
 
+def ooc_cmd_uninvite(client, arg):
```
Comment (oldmud0) [inline server/commands.py:583]: Needs documentation in readme.

File: server/commands.py [hunk 3]
```diff
@@ -527,16 +570,41 @@ def ooc_cmd_invite(client, arg):
...
+    if not client.area.is_locked and not client.is_mod:
+        raise ClientError('Area isn\'t locked.')
+    if not arg:
+        raise ClientError('You must specify a target. Use /uninvite <id>')
+    arg = arg.split(' ')
+    targets = client.server.client_manager.get_targets(client, TargetType.ID, int(arg[0]), True)
+    if targets:
+        try:
+            for c in targets:
+                client.send_host_message("You remove {} from the whitelist.".format(c.get_char_name()))
```
Comment (oldmud0) [inline server/commands.py]: `removed`*

File: server/commands.py [hunk 4]
```diff
@@ -527,16 +570,41 @@ def ooc_cmd_invite(client, arg):
...
+        except AreaError:
+            raise
+        except ClientError:
+            raise
+    else:
+        client.send_host_message("No targets found.")
+
 def ooc_cmd_area_kick(client, arg):
-    if not client.is_cm and not client.is_mod:
+    if not client.is_mod:
```
Comment (oldmud0) [inline server/commands.py:607]: You may want to note this change, that CMs are now allowed to kick players.

File: server/commands.py [hunk 5]
```diff
@@ -181,6 +181,49 @@ def ooc_cmd_pos(client, arg):
...
 
+def ooc_cmd_forcepos(client, arg):
+    if not client.is_cm and not client.is_mod:
+        raise ClientError('You must be authorized to do that.')
+
+    args = arg.split()
+
+    if len(args) < 1:
+        raise ArgumentError(
+            'Not enough arguments. Use /forcepos <pos> <target>. Target should be ID, OOC-name or char-name. Use /getarea for getting info like "[ID] char-name".')
```
Comment (oldmud0) [inline server/commands.py:192]: `pos` does not seem to be checked early for validity.

**Review by oldmud0 — CHANGES_REQUESTED** [review]

File: server/commands.py (hunk 5 above)
Comment (Crystalwarrior) [reply server/commands.py:192]: It's checked by the ClientError bit, actually. Seems to be working out fine.

File: server/commands.py (hunk 4 above)
Comment (Crystalwarrior) [reply server/commands.py:607]: CM's aren't allowed to area-kick players anymore but yeah I'll note that in README.md

File: server/commands.py [hunk 6]
```diff
@@ -181,6 +181,49 @@ def ooc_cmd_pos(client, arg):
...
+
+
+
+    for t in targets:
+        try:
+            t.change_position(pos)
+            t.area.broadcast_evidence_list()
+            t.send_host_message('Forced into /pos {}.'.format(pos))
+        except ClientError:
+            raise
```
Comment (Crystalwarrior) [inline server/commands.py:220]: This is where it brings up the invalid pos error, btw.

**Review by oldmud0 — APPROVED** [review]

**Comment by oldmud0** [comment]
thanks!

**Pull request merged**